hw/tpm/tpm_tis_spi.c

```diff
@@ -0,0 +1,253 @@
+    hwaddr addr;
+    uint8_t bytes[sizeof(hwaddr)];
+};
+
+union TpmTisSpiData {
+    uint32_t data;
+    uint8_t bytes[sizeof(uint32_t)];
+};
+
+struct TPMSpiState {
+    /*< private >*/
+    SSIPeripheral parent_obj;
+
+    /*< public >*/
+    TPMState tpm_state; /* not a QOM object */
+    TisSpiPktState tis_spi_state;
+
+    union TpmTisRWSizeByte first_byte;
+    union TpmTisSpiHwAddr addr;
+    union TpmTisSpiData data;
+
+    uint8_t data_idx;
+    uint8_t addr_idx;
+};
+
+OBJECT_DECLARE_SIMPLE_TYPE(TPMSpiState, TPM_TIS_SPI)
+
+static uint32_t tpm_tis_spi_transfer8(SSIPeripheral *ss, uint32_t tx)
+{
+    TPMSpiState *tts = TPM_TIS_SPI(ss);
+    uint32_t r = 1;
+
+    switch (tts->tis_spi_state) {
+    case TIS_SPI_PKT_STATE_START:
+        tts->first_byte.byte = (uint8_t)tx;
+        tts->first_byte.data_expected_size++;
+        tts->data_idx = 0;
+        tts->addr_idx = TPM_TIS_SPI_ADDR_BYTES;
+        tts->tis_spi_state = TIS_SPI_PKT_STATE_ADDRESS;
+        break;
+    case TIS_SPI_PKT_STATE_ADDRESS:
+        assert(tts->addr_idx > 0);
+        tts->addr.bytes[--tts->addr_idx] = (uint8_t)tx;
+
+        if (tts->addr_idx == 0) {
+            if (tts->first_byte.rwflag == SPI_WRITE) {
+                tts->tis_spi_state = TIS_SPI_PKT_STATE_DATA_WR;
+            } else { /* read and get the data ready */
+                tts->data.data = (uint32_t)tpm_tis_memory_ops.read(
+                                           &tts->tpm_state,
+                                           tts->addr.addr,
+                                           tts->first_byte.data_expected_size);
+                tts->tis_spi_state = TIS_SPI_PKT_STATE_DATA_RD;
+            }
+        }
+        break;
+    case TIS_SPI_PKT_STATE_DATA_WR:
+        tts->data.bytes[tts->data_idx++] = (uint8_t)tx;
+        if (tts->data_idx == tts->first_byte.data_expected_size) {
+             tpm_tis_memory_ops.write(&tts->tpm_state,
+                                      tts->addr.addr,
+                                      tts->data.data,
+                                      tts->first_byte.data_expected_size);
+             tts->tis_spi_state = TIS_SPI_PKT_STATE_START;
+        }
+        break;
+    case TIS_SPI_PKT_STATE_DATA_RD:
+        r = tts->data.bytes[tts->data_idx++];
+        if (tts->data_idx == tts->first_byte.data_expected_size) {
+            tts->tis_spi_state = TIS_SPI_PKT_STATE_DONE;
+        }
+        break;
+    default:
+        tts->tis_spi_state = TIS_SPI_PKT_STATE_START;
+        r = (uint32_t) -1;
+    }
+
+    return r;
+}
+
+static uint32_t tpm_tis_spi_transfer8_ex(SSIPeripheral *ss, uint32_t tx)
+{
+    TPMSpiState *tts = TPM_TIS_SPI(ss);
+    SSIBus *ssibus = (SSIBus *)qdev_get_parent_bus(DEVICE(tts));
+
+    uint8_t prev_state = tts->tis_spi_state;
+    uint32_t r = tpm_tis_spi_transfer8(ss, tx);
+    uint8_t curr_state = tts->tis_spi_state;
+
+    if (ssibus->preread &&
+       /* cmd state has changed into DATA reading state */
```

Comment:
strong suggest add comments here explain our preread hacking. 
give example in my previous comments.